conflation/aggregation.py

```diff
@@ -0,0 +1,139 @@
+        "driveway": None,
+        "alley": None,
+        "parking_aisle": None,
+        "drive-through": None,
+    },
+    "urban": {
+        "way": [None, None, None, None, None, None, None, None],
+        "link_exiting": [None, None, None, None, None],
+        "link_turning": [None, None, None, None, None],
+        "roundabout": [None, None, None, None, None, None, None, None],
+        "driveway": None,
+        "alley": None,
+        "parking_aisle": None,
+        "drive-through": None,
+    },
+}
+ROAD_CLASS_INDEX_MAPPING = {
+    "motorway": 0,
+    "trunk": 1,
+    "primary": 2,
+    "secondary": 3,
+    "tertiary": 4,
+    "unclassified": 5,
+    "residential": 6,
+    "service_other": 7,
+}
+
+
+def run(map_matches_dir: str, results_dir: str) -> None:
+    """
+    This is the final step of the script, where the map matching results from step two are aggregated together to build
+    the config.json that acts as the output of the script. The config.json follows the definition from the
+    OpenStreetMapSpeeds/schema repo.
+
+    It does a walk through the output dirs of the map matching process. Each file in these dirs holds a list of tuples
+    that contain each individual measurement we made during map matching (kind of like database tables). The data is
+    aggregated together with a group_by using pandas.
+
+    :param map_matches_dir: Dir where map match results from step 2 were pickled to
+    :param results_dir: Dir where the final config.json should be stored
+    """
+    # Check to see if the final result has already been processed
+    final_config_filename = util.get_final_config_filename(results_dir)
+    if os.path.exists(final_config_filename):
+        print("Final config already built. Skipping...")
+        return
+
+    final_config = []
+    for subdir, dirs, files in os.walk(map_matches_dir):
+        for file in files:
+            # Pull the country and region using the name of the subdir and the pickle file
+            country = os.path.basename(os.path.normpath(subdir))
+            region = file.split(".")[0]
+
+            # Pull map matches from disk
+            map_matches_filename = os.path.join(subdir, file)
+            try:
+                print("Reading {}/{} map match results from disk...".format(country, region))
+                map_matches: dict[str, dict[str, list[tuple]]] = pickle.load(
+                    open(map_matches_filename, "rb")
+                )
+            except (OSError, IOError):
+                print("ERROR: {} pickle could not be loaded. Cannot perform aggregation.")
+                continue
+
+            df = pd.DataFrame(map_matches, columns=MAP_MATCH_COLS)
+
+            # TODO: For now we're just doing a simple median over the data. Do we need something fancier?
+            final_config.append(
+                measurements_to_config(
+                    df.groupby(MAP_MATCH_COLS[:-1]).median(), country, region
+                )
+            )
```

Comment:
Uses pandas to do a multi-column GROUP BY on ["density", "road_class", "type"], and pulls out the median.